avocado/utils/asset.py

```diff
@@ -71,21 +71,178 @@ def __init__(self, name, asset_hash, algorithm, locations, cache_dirs,
+                self.locations.append(locations)
+        elif isinstance(locations, list):
+            # remove asset name from location
+            for location in locations:
+                if self.basename in location:
+                    parsed_location = urlparse.urlparse(location)
+                    self.locations.append(
+                        os.path.dirname(parsed_location.geturl()))
+                else:
+                    self.locations.append(location)
+
+        # Let's make the user's life easy and handle the cache
+        if cache_dirs:
+            self.cache_dirs = cache_dirs
         else:
-            self.locations = locations
-        self.cache_dirs = cache_dirs
+            # Break early if we don't have the cache_dirs. This is necessary
+            # so we don't change the Asset signature.
+            raise EnvironmentError("At least one cache directory is necessary"
+                                   " to download the asset.")
+
         self.expire = expire
         self.metadata = metadata
 
+        # Break early. We need to know the source location of the asset.
+        if not self.locations:
+            raise EnvironmentError("An asset should have at least one source"
+                                   " location. Double check your asset name"
+                                   " and location and make sure to use a"
+                                   " supported protocol.")
+
+    def _create_hash_file(self, asset_path):
+        """
+        Compute the hash of the asset file and add it to the CHECKSUM
+        file.
+
+        :param asset_file: full path of the asset file.
+        """
+        result = crypto.hash_file(asset_path, algorithm=self.algorithm)
+        with open(self._get_hash_file(asset_path), 'w') as hash_file:
+            hash_file.write('%s %s\n' % (self.algorithm, result))
+
+    def _create_metadata_file(self, asset_dir):
+        """
+        Creates JSON file with metadata.
+        The file will be saved as `asset_file`_metadata.json
+
+        :param asset_dir: full path where the asset file is located.
+        """
+        metadata_file_name = "%s_metadata.json" % self.basename
+        metadata_path = os.path.join(os.path.dirname(asset_dir),
+                                     metadata_file_name)
+        metadata = json.dumps(self.metadata)
+        with open(metadata_path, "w") as metadata_file:
+            metadata_file.write(metadata)
+
+    def _download(self, url_obj, asset_path):
+        """
+        Download the asset from an url.
+
+        :param url_obj: object from urlparse.
+        :param asset_path: full path of the asset file.
+        :returns: if the downloaded file matches the hash.
+        :rtype: bool
+        """
+        try:
+            # Temporary unique name to use while downloading
+            temp = '%s.%s' % (asset_path,
+                              next(tempfile._get_candidate_names()))  # pylint: disable=W0212
```

Comment:
This is very weird. I look at `tempfile` documentation and I did not find a way to get the candidate name only.
I did a quick experiment (and I might have missed something!):

```
temp = tempfile.NamedTemporaryFile()
url_download(url_obj.geturl(), temp.name)
```
Even holding a reference to temporary file on `temp` the file can be opened inside `url_download`. But I think you will need to copy `temp.name` over `asset_path` (instead of move - `shutil.move(temp, asset_path))` in the `FileLock`.  Ah, you also need to `temp.close()` after copy.

Well, perhaps not too simple but...

Reply:
This is a part of the code that I'm not happy with. Accessing a private method from `tempfile` does not look like a good idea. I left the code as it is in the original class to be able to discuss it here.

One of the ideas I had was to generate a random number and add it to the temporary file name. We cannot rely on user's `/tmp` space and use a temporary file as the asset may not fit there. That may be one of the reasons the original download method uses the Avocado cache.

Related to the `copy`/`move` of the file, the original method was downloading, making a copy, then deleting the original file. This caused the use of double the space of the asset and generate an overhead with the `copy` that is not necessary, that is why I changed it to a download then move. So far I have not seen a side effect.